Fix curve color cycle mode

markcmiller86 opened the pull request:
### Description

* Resolves #5067 
* Returns much of the logic to its state pre- 2acd674b23ae43a89f563ebc9aac9c3f6470186b where autonomous color mode selection logic was in PluginInfo routines and not in the Plot object itself.
* But, also simplifies that logic slightly be replacing the custom `XXXViewerEnginePluginInfo::InitializeGlobalObjects()` with the builtin version and moving `colorIndex` static variable and initialization to `XXXViewerEnginePluginInfo::SetAutonomousColors()`
* Also improves behavior and consistency-ifies similar feature in Mesh plot.
* Adds method, `avtColorTables::ModifyColor(...)` to lighten or darken an rgb color by a multipilicative factor.
* Adds foreground and background colors to `avtPlotMetaData` so that newly created plots that have auto-color selection options can be informed about what current foreground and background colors are.

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have updated the release notes
- [ ] I have made corresponding changes to the documentation
- [ ] I have added debugging support to my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added any new baselines to the repo
- [x] I have assigned reviewers (see [VisIt's PR procedures](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers) for more information).

markcmiller86 committed:
add method to lighten/darken a color

markcmiller86 committed:
Merge branch '3.1RC' into bug-mcm86-28oct20-curve-color-cycling

markcmiller86 commented:
We appear to use a variety of types for storing color information...

* As `double[3]` (so no alpha) as in https://github.com/visit-dav/visit/blob/6795eb4737175c8c74937cb1ec103f0c9605ab8d/src/viewer/core/ViewerPlot.h#L563
* As `double[4]` (so with alpha) as in https://github.com/visit-dav/visit/blob/6795eb4737175c8c74937cb1ec103f0c9605ab8d/src/avt/Plotter/avtVariableLegend.h#L143
* as `unsigned char[3]` as in https://github.com/visit-dav/visit/blob/6795eb4737175c8c74937cb1ec103f0c9605ab8d/src/avt/Plotter/avtVectorGlyphMapper.h#L110
* and `unsigned char[4]` as in https://github.com/visit-dav/visit/blob/6795eb4737175c8c74937cb1ec103f0c9605ab8d/src/common/state/ColorAttribute.h#L104
* and the above types are re-packaged into things like `ColorAttribute` and `ColorControlPoint` where the only difference between these two packagings of color is that one has an assocaited *position* (control point) and the other doesn't.

markcmiller86 committed:
add bgColor

markcmiller86 committed:
add bgColor

markcmiller86 committed:
move color selection logic to attrib init

markcmiller86 committed:
remove color selection logic

markcmiller86 committed:
add fg/bg colors to avtPlotMetaData

markcmiller86 committed:
adjust color selection to attrib initialization

markcmiller86 committed:
Merge branch 'bug-mcm86-28oct20-curve-color-cycling' of github.com:visit-dav/visit into bug-mcm86-28oct20-curve-color-cycling

markcmiller86 committed:
improve initilization logic

markcmiller86 committed:
remove auto color selection logic

markcmiller86 committed:
add auto color selection logic similar to Curves

markcmiller86 committed:
ensure regen works

markcmiller86 committed:
update relnotes

biagas reviewed (commented):
I haven't actually looked at code yet.
I checked out your branch and compiled and performed a few tests.
I created a Pseudocolor plot of dbA00.pdb, then performed two lineouts.
They both received the same color, even though the default color mode for curve is 'Cycle'.
Also, the color for the refline drawn on the original plot doesn't match the color used for the lineout curve.

brugger1 commented on src/resources/help/en_US/relnotes3.1.4.html:
```diff
@@ -36,6 +36,7 @@
   <li>Fixed a bug where applying Log scaling to the view of a Curve plot (or lineout curve) yielded wrong results.</li>
   <li>Fixed a bug with the variable menu not being sorted properly when the variable list had mixed-case names and variable grouping was used.</li>
   <li>Fixed a bug with the conn_cmfe expression where mapping a variable from a polyhedral mesh to a point mesh failed when the mesh was material selected.</li>
+  <li>Fixed bug (<a href="https://github.com/visit-dav/visit/issues/5067">5067</a>) causing Curve plots with cycle coloring option to change their color as timestep is advanced and a similar issue for random coloring of Mesh plots.</li>
```
I would change the release notes after the link to:

causing Curve plots with the "Curve color" set to "Cycle" to change their color as the timestep is advanced and a similar issue with the Mesh plot when the "Mesh color" or "Opaque color" was set to "Random".

brugger1 reviewed (changes requested):
I tried the curve fill color option with curve plots with 3.1.1 just to understand how fill color1 and fill color2 were used and I noticed that the curve filling was broken. Were there changes in this commit to address this as well? You don't need to fix it if you didn't. I was just curious since you changed how the lighter color was set.

markcmiller86 commented:
> I haven't actually looked at code yet.
> I checked out your branch and compiled and performed a few tests.
> I created a Pseudocolor plot of dbA00.pdb, then performed two lineouts.
> They both received the same color, even though the default color mode for curve is 'Cycle'.
> Also, the color for the refline drawn on the original plot doesn't match the color used for the lineout curve.

I performed these same tests on my branch and it all works fine. I think you are somehow running the old code.

markcmiller86 commented:
> Were there changes in this commit to address this as well?

No.

markcmiller86 commented:
> I was just curious since you changed how the lighter color was set.

I actually reverted to the logic that existed pre- 2acd674b23ae43a89f563ebc9aac9c3f6470186b except with some improvements in how ligthening is handled.

biagas commented:
> 
> 
> > I haven't actually looked at code yet.
> > I checked out your branch and compiled and performed a few tests.
> > I created a Pseudocolor plot of dbA00.pdb, then performed two lineouts.
> > They both received the same color, even though the default color mode for curve is 'Cycle'.
> > Also, the color for the refline drawn on the original plot doesn't match the color used for the lineout curve.
> 
> I performed these same tests on my branch and it all works fine. I think you are somehow running the old code.

My bad. Somehow I had checked out the wrong branch. Sorry about that.

markcmiller86 commented:
I am seeing two issues that are worth fixing in this PR.

1. The lightened color is too often white or nearly so. I think reducing the mult-factor I used may be sufficient
2. The GUI does not reflect the autonomous color selection choice. This issue goes back to at least before version 2.10. I think when the autonomous color selection logic triggers, it needs to somehow inform the GUI what color it selected for the curve. @biagas...any idea where I should look to correct?

brugger1 commented:
I saw the comment about reverting changes, but didn't look at what they were. I'll take a look now to better understand these changes.

markcmiller86 committed:
revise as per @brugger1

biagas commented:
> 
> 
> I am seeing two issues that are worth fixing in this PR.
> 
>     1. The lightened color is too often white or nearly so. I think reducing the mult-factor I used may be sufficient
> 
>     2. The GUI does not reflect the autonomous color selection choice. This issue goes back to at least before version 2.10. I think when the autonomous color selection logic triggers, it needs to somehow inform the GUI what color it selected for the curve. @biagas...any idea where I should look to correct?

Are you saying that the color button for 'Custom' should be updated to reflect the actual color being used when 'Cycle' is selected?
Then perhaps the following section in QvisCurvePlotWindow::UdpateWindow should have the if (curveColor->isEnabled) test removed.

```
        case CurveAttributes::ID_curveColor:
            { // new scope
              if (curveColor->isEnabled())
              {
                QColor temp(atts->GetCurveColor().Red(),
                            atts->GetCurveColor().Green(),
                            atts->GetCurveColor().Blue());
                curveColor->blockSignals(true);
                curveColor->setButtonColor(temp);
                curveColor->blockSignals(false);
              }
            }

```

brugger1 commented:
I originally tried the curve fill color on Windows and 2.13.2 worked and 3.1.1 failed. I went onto quartz and 2.13.3 works (it appears to have bugs, but it sorta works) and with 3.0.1 and later, the fill has no impact. Since you're changing the fill color in response to it being to white, I'm assuming it works for you. Is that correct?

markcmiller86 commented:
> I'm assuming it works for you. Is that correct?

Hadn't tested fill behavior. The code you see here is my attempt to return it to the state it was in pre- 2acd674b23ae43a89f563ebc9aac9c3f6470186b where it had logic to pick a lighter color for the fill2 color. I will test now.

markcmiller86 commented:
> Are you saying that the color button for 'Custom' should be updated to reflect the actual color being used when 'Cycle' is selected?

Yeah, that is what I mean. Thanks for the pointer. Will try there first. I think I need a clientAtts->Notify() from the logic to pick the color in CurveViewerEnginePluginInfo::SetAutonomousColor() too though.

brugger1 commented:
The old code looked simpler and was inlined rather than calling a function. So I assume you are restoring the code and improving it at the same time.

markcmiller86 committed:
adjust to ensure GUI reflects auto color choice

markcmiller86 committed:
adjust to ensure GUI reflects auto color choice

biagas commented:
@markcmiller86 Do you still have logic in there to change the color if it matches the background?
Change background to red (first cycle color)
Open curve.visit, create a plot. The curve color isn't adjusted for the background, its red. And yes, I've verified I am working from the latest version of your branch this time.

markcmiller86 commented:
> Do you still have logic in there to change the color if it matches the background?

Yes but I assumed incorrectly that ViewerPlot's `SetBackgroundColor()` would be called whenever that was changed and its not. In fact, I have no idea what `ViewerPlot::SetBackgroundColor()` is there for if not to be called whenever the background color is changed.

markcmiller86 committed:
adjust where to get bg/fg color info

markcmiller86 committed:
add SetClientAtts call so GUI reflects color choice

markcmiller86 committed:
ensure sqrt arg is handled as double

markcmiller86 committed:
use 1.75 as lightening factor

markcmiller86 committed:
Merge branch '3.1RC' into bug-mcm86-28oct20-curve-color-cycling

markcmiller86 committed:
fix comment

markcmiller86 committed:
ensure GUI reflects any auto color choices

markcmiller86 committed:
Merge branch 'bug-mcm86-28oct20-curve-color-cycling' of github.com:visit-dav/visit into bug-mcm86-28oct20-curve-color-cycling

markcmiller86 committed:
fix comment

markcmiller86 commented:
@brugger1 and @biagas, I think this is ready for re-review...
* I have fixed [issues I mentioned](https://github.com/visit-dav/visit/pull/5205#issuecomment-721405266)
* I corrected the relnotes comment @brugger1 requested changed
* I fixed issue with de-conflicting selected color from background color [@biagas mentioned](https://github.com/visit-dav/visit/pull/5205#issuecomment-721460973)

markcmiller86 commented:
BTW...don't be confused. I just a few mins. ago closed a 6 month old PR I had started to address this and then let languish.

markcmiller86 commented:
Oh, and @biagas, I forgot to ask...In this PR, should I change name of equiv. functionality in Mesh plot to use `Cycle` verbiage instead of `Random` there? I can't recall what my original thinking was then except that its really effectively a random color choice.

markcmiller86 committed:
Merge branch '3.1RC' into bug-mcm86-28oct20-curve-color-cycling

biagas commented on src/viewer/core/ViewerPlot.C:
```diff
@@ -4687,6 +4706,8 @@ ViewerPlot::UpdateColorTable(const std::string &ctName)
 bool
 ViewerPlot::SetBackgroundColor(const double *bg)
 {
+printf("ViewerPlot::SetBackgroundColor = %f, %f, %f\n",
```
Need to remove the printf.

biagas commented on src/avt/Pipeline/Data/avtColorTables.C:
```diff
@@ -508,6 +508,53 @@ reverse_alphas(unsigned char *a, int na)
+//
+// Purpose: Lighten or darken a color by a multiplicative factor without
+//          changing the effective hue. To lighten a color pass a value for
+//          mult > 1 but not to much larger than 2. To darken a color pass a
```
'too much' instead of 'to much'

biagas commented on src/viewer/core/ViewerPlot.C:
```diff
@@ -1612,9 +1615,25 @@ ViewerPlot::GetPlotMetaData() const
     }
+
+    /* Get bg/fg color info from window. Even though this-> object
+       appears to have it also, it is not reliably kept up to date. */
```
This is unfortunate. Can you enter a ticket to investigate why the background and foreground colors aren't sent to the ViewerPlot reliably?

biagas reviewed (changes requested):
Thanks for fixing the matches-with-background adjustment. There is still a slight problem.
After a curve color gets adjusted in this fashion, the very next curve added uses the same color.

markcmiller86 committed:
remove printf

markcmiller86 committed:
fix typo

markcmiller86 committed:
adjust interface for GetJNDControlPointColor

markcmiller86 committed:
remove printf

markcmiller86 committed:
skip n colors attempted

markcmiller86 committed:
fix use of fg/bg for mesh/opaque cases; skip n colors attempted

markcmiller86 committed:
Merge branch 'bug-mcm86-28oct20-curve-color-cycling' of github.com:visit-dav/visit into bug-mcm86-28oct20-curve-color-cycling

markcmiller86 commented:
> After a curve color gets adjusted in this fashion, the very next curve added uses the same color.

Oh, thats right. I had fixed this on another branch. I adjusted the logic in `avtColorTables::GetJNDControlPointColor` to return the number of colors it skipped before finding an acceptable one and then update the `colorIndex` used in the plot's ViewerEngineInfo objects by that amount.

biagas commented:
Are lineout generated curves within the purview of this PR?
Open a 2d dataset and create a pseudocolor plot.
Open a new window and change its background color to red.
Do a lineout in the first window
the lineout curve gets plotted in the second windows, but is red, and disappears.

markcmiller86 commented:
> Are lineout generated curves within the purview of this PR?

Possibley.

> Open a 2d dataset and create a pseudocolor plot.
> Open a new window and change its background color to red.
> Do a lineout in the first window
> the lineout curve gets plotted in the second windows, but is red, and disappears.

The lineout curve *should* be getting the background color of the window in which it is being displayed. Let me check what is going on in this case.

markcmiller86 commented:
When (and where in the code) is a *lineout* curve's color selected? I think a lineout *line* color and lineout curve color are bound together and I think the curve's color is determined *after* the line color in the first window. There, the selection of red is fine because the background is white. We'd have to make the lineout line color selection logic use the background color of the lineout curve window as the color to de-conflict with.

markcmiller86 commented:
So, I am seeing some oddities w.r.t. how the **Curve Plot** attributes associated with a lineout are handled.

- Default Curve attributes are not obeyed
  - Open rect2d.silo
  - Select Curve from Plot Attributes pull down
    - Set to custom color, yellow
    - Make default
  - PC plot d and then do a lineout of it
    - you get a red curve but with color set to custom, not cycle.
- It looks like the code used to determine the line color is https://github.com/visit-dav/visit/blob/a72c57ed3347979de9b3e0e61c9857c3aec98893/src/viewer/core/ViewerQueryManager.C#L2940-L2963
- It would be best if VQM called the color selection code in `CurveViewerEnginePluginInfo::SetAutonomousColor()`

markcmiller86 commented:
I think the solution to address [this request](https://github.com/visit-dav/visit/pull/5205#issuecomment-722038506) is to move the color selection logic *out* of `XXXViewerEnginePluginInfo.C` to `avtColorTables.C` where it will then be available to `ViewerQueryManager.C`, `CurveViewerEnginePluginInfo.C` and `MeshViewerEnginePluginInfo.C`.

markcmiller86 committed:
track color index using map<string,int>

markcmiller86 committed:
Track color index via avtColorTables

markcmiller86 committed:
adjust lineout color selection logic

markcmiller86 commented:
Ok, I've got the lineout color de-conflicted with the lineout window background now.

Previously, every caller that wanted to march through discrete colors picking a *next* one (but avoiding one too close to another color like background), maintained their own `static int colorIndex`. I've removed all that and now track those in `avtColorTables` in a `static map<string, int>` there. That seems to work ok. I've haven't finished testing though. And, I need to do one more fix w.r.t. to mesh coloring for opaque and lines. Opaque should be de-conflicted with background and then lines deconflicted with opaque color. Thats just a re-order of the existing ops though.

markcmiller86 committed:
deconflict mesh (lines) with opaque color

markcmiller86 committed:
Merge branch '3.1RC' into bug-mcm86-28oct20-curve-color-cycling

markcmiller86 committed:
Merge branch 'bug-mcm86-28oct20-curve-color-cycling' of github.com:visit-dav/visit into bug-mcm86-28oct20-curve-color-cycling

biagas reviewed (approved):
Great job working through all the issues raised!
